src/arcsf/data/data_utils.py

```diff
@@ -0,0 +1,244 @@
+        )
+    )
+
+    return forget_authors
+
+
+def load_tofu(
+    granularity: str = "author_level",
+    forgotten_author_fraction: float = 0.1,
+    forgotten_fact_fraction: float = 0.1,
+    random_seed: int = 42,
+    debug=False,
+):
+    """
+    Loads the TOFU dataset with randomly chosen authors/information to forget,
+    using question indices to remove/retain data.
+
+        Parameters:
+                granularity (str):                          Granularity at which to
+                                                            perform unlearning
+                forgotten_author_fraction (float):          Fraction of authors that
+                                                            should be removed
+                forgotten_fact_fraction (float):            Fraction of facts to be
+                                                            removed (currently not used)
+                random_seed (int):                          Random seed for
+                                                            reproducibility
+
+        Returns:
+                forget_set (datasets.arrow_dataset.Dataset):    Dataset containing
+                                                                removed/forgotten
+                                                                authors
+                retain_set (datasets.arrow_dataset.Dataset):    Dataset containing
+                                                                retained authors
+                debug_dict (dict):                              Dictionary containing
+                                                                information for
+                                                                debugging, used in tests
+    """
+
+    all_data = load_dataset("locuslab/TOFU", "full")[
+        "train"
+    ]  # load all data to work with
+    author_count = 200  # hard coding author count for now
+    author_q_count = 20  # hard coding author question count for now
+    qs_to_remove = 4
+    debug_dict = {"author_count": author_count, "author_q_count": author_q_count}
+
+    forget_author_numbers = get_author_splits(
+        author_count, forgotten_author_fraction, random_seed
+    )
+    debug_dict["forget_author_numbers"] = forget_author_numbers
+
+    if granularity == "author_level":
+        forget_indices, retain_indices = get_indices_structured(
+            forget_author_numbers,
+            q_remove=author_q_count,
+            n_authors=author_count,
+            q_per_author=author_q_count,
+        )
+
+    elif granularity == "structured_within_authors":
+        # biographical information generally contained in first 4 questions
+        forget_indices, retain_indices = get_indices_structured(
+            forget_author_numbers,
+            q_remove=qs_to_remove,
+            n_authors=author_count,
+            q_per_author=author_q_count,
+        )
+        debug_dict["qs_to_remove"] = qs_to_remove
```

Comment:
I've been trying to think about how to reduce this function in length - it feels like it could be shorter!

Generating the full set of indices is fairly straightforward:
```python
all_indices = list(range(all_data.num_rows))
```

The awkward thing is trying to generate the forget set conditional on the three (see comment above) kwargs for determining how to handle this.

The four scenarios are:

1. Sample x authors to remove all questions for
2. Stratified
    a. Remove the first x questions within authors (with a view to normally removing 4, but there's no need to hardcode this)
    b. Randomly remove x questions within authors
3. Non-stratified: just remove x questions at random

2b makes this a little trickier (because the ordering is semi-deterministic for type of question, we should randomly sample questions for each author). I'd suggest using a function to handle generating the question numbers as a result. Where `granularity`, `stratify`, `forget_random`, and `num_forget` are all kwargs to `load_tofu`:

```python
def make_question_generator(forget_random):
    def generate_questions(q_per_author, num_forget):
        if forget_random:
            return random.sample(range(q_per_author), num_forget)
        if not forget_random:
            return range(list(num_forget))
    return generate_questions

# Determine list of authors to consider forgetting based on granularity
if granularity == "question":
    if not stratify:
        forget_indices = random.sample(all_indices, num_forget) #job done!
    if stratify:
        authors_to_forget = list(range(num_authors))
        generate_questions = make_question_generator(forget_random)
if granularity == "author":
    authors_to_forget = random.sample(range(num_authors), num_forget)
    num_forget = q_per_author #set to this value for further forgetting
    questions = generate_questions = make_question_generator(False)

# Generate indices based on selections
if stratify or granularity == "author":
    forget_indices = sum([
        get_forget_indices(
            author,
            generate_questions(q_per_author, num_forget),
            q_per_author,
        ) for
        author in authors_to_forget
    ], []) #sum is a hacky way of flattening
```

There may be a way to further simplify this but it's not occurring to me at the moment.

You can get the final sets of indicies by:

```python
retain_indices = list(set(all_indices).difference(forget_indices))
```

I think at some point it's worth having a conversation about whether we want to include the deterministic dropping (i.e. drop first x within dataset)

Reply:
Okay, ignore the above - I shouldn't write code recommendations near to 5pm. I'd suggest the following, using train_test_split from scikit-learn:

For getting forget numbers from question/author number:

```python
def flatten_list_of_lists(
    list: list[list]
) -> list:
    return sum(list, [])

def _get_forget_index(
    author: int,
    question: int,
    q_per_author: int,
) -> int:
    return question + q_per_author * author

def _get_forget_indices(
    authors: int,
    questions: int | list[int],
    q_per_author: int,
) -> int | list[int]:
    if isinstance(questions, int):
        return _get_forget_index(authors, questions, q_per_author)
    if isinstance(questions, list):
        return [_get_forget_index(authors, question, q_per_author) for question in questions]

def get_forget_indices(
    authors: int | list[int],
    questions: int | list[int],
    q_per_author: int,
) -> list[int]:
    if isinstance(authors, int):
        return _get_forget_indices(authors, questions, q_per_author)
    if isinstance(authors, list):
        out = [_get_forget_indices(author, questions, q_per_author) for author in authors]
        if isinstance(out[0], list):
            return flatten_list_of_lists(out)
        return out
```

For handling sampling:
```python
# Get full list of indices
all_indices = list(range(all_data.num_rows))

# Get full list of authors
authors = list(range(num_authors))

# Get indices
if granularity == "question":
    if forget_random:
        retrain_indices, forget_indices = train_test_split(
            all_indices,
            test_size=num_forget,
            stratify=sum([[a]*num_authors for a in authors], []) if stratify
            else None,
            random_state=random_seed,
        )
    if not forget_random:
        questions = list(range(q_per_author))
        forget_indices = get_forget_indices(authors, questions, q_per_author)
        retain_indices = list(set(all_indices).difference(forget_indices))
if granularity == "author":
    retain_indices, forget_indices = train_test_split(
        authors,
        test_size=num_forget,
        random_state=random_seed,
    )
```

In general - if an existing function already does what you need, it's best to use this rather than write your own code (I should've started here yesterday!). The existing implementation will both be better than your own (as it'll have gone through far more rounds of development/more people) and makes life easier.

The nice thing about `train_test_split` is the number of questions to drop doesn't need to be a multiple of the number of authors - we can drop 7 questions and it will simply ensure only one question from each author is dropped at most for a total of 7 (whereas in my code from yesterday, we could only choose to drop x questions per author).